checks.d/consul.py

```diff
@@ -0,0 +1,215 @@
+
+    def get_nodes_with_service(self, instance, service, tag=None):
+        if tag:
+            consul_request_url = '/v1/catalog/service/{0}?tag={1}'.format(service,tag)
+        else:
+            consul_request_url = '/v1/catalog/service/{0}'.format(service)
+
+        return self.consul_request(instance, consul_request_url)
+
+    def get_services_on_node(self, instance, node):
+        return self.consul_request(instance, '/v1/catalog/node/{0}'.format(node))
+
+    def check(self, instance):
+        perform_new_leader_checks = instance.get('new_leader_checks',
+                                                 self.init_config.get('new_leader_checks', False))
+        if perform_new_leader_checks:
+            self._check_for_leader_change(instance)
+
+        if not self._is_instance_leader(instance):
+            self.log.debug("This consul agent is not the cluster leader." +
+                           "Skipping service and catalog checks for this instance")
+            return
+
+        service_whitelist = instance.get('service_whitelist',
+                                         self.init_config.get('service_whitelist', []))
+
+        service_check_tags = ['consul_url:{0}'.format(instance.get('url'))]
+        perform_catalog_checks = instance.get('catalog_checks',
+                                              self.init_config.get('catalog_checks'))
+
+        try:
+            health_state = self.consul_request(instance, '/v1/health/state/any')
+
+            STATUS_SC = {
+                'passing': AgentCheck.OK,
+                'warning': AgentCheck.WARNING,
+                'critical': AgentCheck.CRITICAL,
+            }
+
+            for check in health_state:
+                status = STATUS_SC.get(check['Status'])
+                if status is None:
+                    continue
+
+                tags = ["check:{0}".format(check["CheckID"])]
+                if check["ServiceName"]:
+                    tags.append("service:{0}".format(check["ServiceName"]))
+                if check["ServiceID"]:
+                    tags.append("service-id:{0}".format(check["ServiceID"]))
+
+            services = self.get_services_in_cluster(instance)
+            self.service_check(self.HEALTH_CHECK, status, tags=tags)
+
+        except Exception as e:
+            self.service_check(self.CONSUL_CHECK, AgentCheck.CRITICAL,
+                               tags=service_check_tags)
+        else:
+            self.service_check(self.CONSUL_CHECK, AgentCheck.OK,
+                               tags=service_check_tags)
+
+        if perform_catalog_checks:
+            services = self.get_services_in_cluster(instance)
+            if service_whitelist:
+                services = [s for s in services if s in service_whitelist]
+            else:
+                #Default to polling all services
+                self.log.warn('Consul service whitelist not defined. Agent will poll for all %d services found' % len(services))
```

Comment:
Also we should limit the number of services to pull, you don't want to pull 1000 services.